apps/backend/src/modules/rating/controller.ts

```diff
@@ -0,0 +1,501 @@
+import { GraphQLError } from "graphql";
+import { connection } from "mongoose";
+
+import { AggregatedMetricsModel, RatingModel } from "@repo/common";
+import { METRIC_MAPPINGS } from "@repo/shared";
+
+import { MetricName, Semester } from "../../generated-types/graphql";
+import {
+  formatAggregatedRatings,
+  formatSemesterRatings,
+  formatUserClassRatings,
+  formatUserRatings,
+} from "./formatter";
+import {
+  courseRatingAggregator,
+  ratingAggregator,
+  semestersWithRatingsAggregator,
+  userClassRatingsAggregator,
+  userRatingsAggregator,
+} from "./helper/aggregator";
+import {
+  checkRatingExists,
+  checkUserMaxRatingsContraint,
+  checkValueConstraint,
+} from "./helper/checkConstraints";
+
+export const numberScaleMetrics = Object.entries(METRIC_MAPPINGS)
+  .filter(([_, config]) => config.isRating)
+  .map(([metric]) => metric) as MetricName[];
+
+export const booleanScaleMetrics = Object.entries(METRIC_MAPPINGS)
+  .filter(([_, config]) => !config.isRating)
+  .map(([metric]) => metric) as MetricName[];
+
+// const getSemestersByInstructor = async (
+//   professorName: string,
+//   subject: string,
+//   courseNumber: string
+// ): Promise<string[]> => {
+//   const semesterInstances = await semestersByInstructorAggregator(
+//     professorName,
+//     subject,
+//     courseNumber
+//   );
+//   const result = semesterInstances.map((i: any) => i._id);
+//   return result;
+// };
+
+// not fully implemented
+// export const getRatingByInstructor = async (
+//   professorName: string,
+//   subject: string,
+//   courseNumber: string
+// ) => {
+//   const semesters = await getSemestersByInstructor(
+//     professorName,
+//     subject,
+//     courseNumber
+//   );
+//   return semesters;
+// };
+
+export const createRating = async (
+  context: any,
+  subject: string,
+  courseNumber: string,
+  semester: Semester,
+  year: number,
+  classNumber: string,
+  metricName: MetricName,
+  value: number
```

Comment:
`userId: string`

let's order the parameters from most general to most specific (do this for all functions please):
1. year
2. semester
3. subject
4. courseNumber
5. classNumber
6. metricName
7. value